kube-runtime/src/observer.rs

```diff
@@ -0,0 +1,114 @@
+use backoff::ExponentialBackoff;
+use futures::{future::ready, Stream, StreamExt};
+use kube_client::{
+    api::{Api, ListParams},
+    core::Resource,
+};
+use serde::de::DeserializeOwned;
+use std::fmt::Debug;
+
+use crate::{
+    utils,
+    watcher::{watcher, Error, Event, Result},
+};
+
+
+/// A simple observer around a watcher with error handling and retry backoff
+///
+/// # Error handling
+///
+/// An `Observer` sets a sensible default backoff policy for all watch events and will retry
+/// (with expotential backoff) from transient `watcher` failures until the retry policy is breached.
+///
+/// To configure the retry policy use `Observer::backoff`.
+///
+/// Note that while all errors will be retried, some represent the need for user action to recover:
+///
+/// - 404 `ErrorResponse`(watching invalid / missing api kind/group for `K`)
+/// - 403 `ErrorResponse` (missing list + watch rbac verbs for `K`)
+pub struct Observer<K>
+where
+    K: Clone + Resource + Send + Sync + 'static,
+{
+    // temporary builder params
+    api: Api<K>,
+    listparams: Option<ListParams>,
+    backoff: Option<ExponentialBackoff>,
+}
+
+impl<K> From<Api<K>> for Observer<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+{
+    fn from(api: Api<K>) -> Observer<K> {
+        Observer::new(api)
+    }
+}
```

Comment:
Not quite sure about this, doesn't feel like `Api` and `Observer` are trying to model the same concept. Additionally, the implied `Into` instance would basically never be useful.

Reply:
good point. removing.